### PR TITLE
use int64 for samples instead of bool8

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
The new qubit device was going to return booleans for 0-1 samples because that is what a boolean represents. However, this change is different from what we had before.~, and booleans still take a full byte of memory to stay aligned (let's not get too crazy here in Pythonland anyway) which is the same as `np.int8`.~

**Description of the Change:**
`sample_state`, and consequentially `qml.sample()` now returns samples of dtype=np.int64 instead of np.bool8.

**Benefits:**
- The user will receive the more-familiar int return type.
- There are no problems with memory-alignment
- The new device API is consistent with the old one
- existing interface code still works without change

**Possible Drawbacks:**
`np.bool8` was smaller

### CODE REVIEW

File: pennylane/measurements/sample.py [hunk 1]
```diff
@@ -140,15 +140,15 @@ def return_type(self):
     @property
     @functools.lru_cache()
     def numeric_type(self):
-        # Note: we only assume an integer numeric type if the observable is a
+        # Note: we only assume a boolean (0-1) numeric type if the observable is a
         # built-in observable with integer eigenvalues or a tensor product thereof
         if self.obs is None:
             # Computational basis samples
-            return int
+            return bool
```
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/measurements/sample.py:147]: to distinguish from ClassicalShadowMP.numeric_type

File: pennylane/devices/qubit/sampling.py [hunk 2]
```diff
@@ -349,4 +349,5 @@ def sample_state(
         samples = rng.choice(basis_states, shots, p=probs)
 
     powers_of_two = 1 << np.arange(num_wires, dtype=np.int64)[::-1]
-    return (samples[..., None] & powers_of_two).astype(np.bool8)
+    states_sampled_base_ten = samples[..., None] & powers_of_two
+    return (states_sampled_base_ten > 0).astype(np.int64)
```
Comment (mudit2812) [inline pennylane/devices/qubit/sampling.py:353]: Why not `np.int8`?

Comment (timmysilv) [reply pennylane/devices/qubit/sampling.py:353]: updated the PR details a bit, but as you can see in [this CI run](https://github.com/PennyLaneAI/pennylane/actions/runs/6013291622/job/16310476417) from before I switched from int8 to int64, the _old_ device was jitted thinking it was going to get int8, but then it got int64. This is because the conversion from `tape.numeric_type == int <--> jit_type=jnp.int8` happens in one place, which should ideally be device-agnostic.

I _could_ start making device-specific changes/branches in our code, but that would become messy. I could also change the old device API to _also_ use `np.int8`, but I'm worried that some plugins/other devices might have sample-dtype overrides by accident, and then it'll clash with our interface code.

tl;dr it's just not worth the effort rn, we're not low on space

Comment (mudit2812) [reply pennylane/devices/qubit/sampling.py:353]: That sounds reasonable. I think it's worth making a story for this though. Using 8 times more memory than "necessary" shouldn't be around long term.

Comment (albi3ro) [reply pennylane/devices/qubit/sampling.py:353]: We are all on board with you there.

**Review by albi3ro — APPROVED** [review]
Sad that we have to do this, but it allows us to move forward.
